### PR TITLE
Add action to verify PR head and base branches

### PR DESCRIPTION
**Pull request by labkey-tchad**
#### Rationale
People occasionally open pull requests against the wrong branch. The `validate-pr` action checks that new pull requests are targeting the correct branch based on the feature branch name (e.g. `fb_*` -> `develop` and `20.11_fb_*` -> `release20.11-SNAPSHOT`). It also warns against using the default PR title generated by GitHub.

There are a couple of bugs in the existing `branch-release` action.
1. It fails if we have already merged changes forward independently of the maintenance release cycle
2. If there are conflicts from the merge forward, we create a placeholder branch and PR to allow somebody to manually resolve the conflicts. TeamCity status checks run and fail on the placeholder branch, often blocking the merge of the maintenance release.

#### Changes
* Add `validate-pr` action
* Update `branch-release` action to not fail if changes have already been merged forward
* Update `branch-release` action to insert an empty commit in merge conflict placeholder branch

### CODE REVIEW

**Comment by labkey-klum** [comment]
How does adding an empty commit message work around TeamCity running and potentially failing a status check on a merge conflict?

**Review by labkey-klum — APPROVED** [review]

File: validate-pr/validate_pr.sh [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+} >>"$ERROR_FILE"
+
+DEFAULT_TITLE="^Fb | fb "
+# Warn about using default pull request title
+if [[ "$PR_TITLE" =~ $DEFAULT_TITLE ]]; then
+    if [ "$(wc -l <"${ERROR_FILE}")" -gt 0 ]; then
+        echo "" >>"$ERROR_FILE" # Insert blank line after other errors.
+    fi
+    echo "This PR appears to have the default title generated by GitHub. Please use something more descriptive." >>"$ERROR_FILE"
+fi
```
Comment (curiositycasualty) [inline validate-pr/validate_pr.sh:77]: Seems like this check could occur at the start of the script-- although I don't see any `git` or `hub` use in the earlier logic, so moving this probably won't make for time savings.

File: validate-pr/validate_pr.sh [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+if [ "$(wc -l <"${ERROR_FILE}")" -gt 0 ]; then
+    cat "$ERROR_FILE" >&2
+    if [ "$PR_NUMBER" != "TEST" ]; then
+        hub api "repos/${GITHUB_REPOSITORY}/issues/${PR_NUMBER}/comments" --field "body=@${ERROR_FILE}"
```
Comment (curiositycasualty) [inline validate-pr/validate_pr.sh:82]: There's not a lot of protection or detection for if this command fails.

**Review by curiositycasualty — APPROVED** [review]
Overall looks good. Just some minor considerations that need not block release.

File: validate-pr/validate_pr.sh [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+SNAP_PATTERN='^release([0-9]+\.[0-9]+)-SNAPSHOT$' # release20.11-SNAPSHOT
+FF_PATTERN='^ff_([0-9]+\.[0-9]+).+'               # ff_20.11.0
+FB_PATTERN='^fb_.+'                               # fb_newFeature_12345
+RFB_PATTERN='^([0-9]+\.[0-9]+)_fb_.+'             # 20.11_fb_backportFeature_12345
+
+{
+    # 'fb_*' should target 'develop'
+    if [[ "$HEAD_BRANCH" =~ $FB_PATTERN ]]; then
+        if [ "$BASE_BRANCH" != "develop" ]; then
+            echo "Expect PR from \`${HEAD_BRANCH}\` to target \`develop\`, not \`${BASE_BRANCH}\`"
```
Comment (curiositycasualty) [inline validate-pr/validate_pr.sh]: If I ever have to do anything more than twice, I usually turn that into a function. So all the instances of this error message could look like: `unexpected_pr_target "${HEAD_BRANCH}" 'develop' "${BASE_BRANCH}"`

**Pull request merged**